pytential/symbolic/execution.py

```diff
@@ -257,6 +257,10 @@ def map_interpolation(self, expr):
         else:
             raise TypeError("cannot interpolate `{}`".format(type(operand)))
 
+    def map_flatten(self, expr):
+        from pytential.utils import flatten_if_needed
+        return flatten_if_needed(self.array_context, self.rec(expr.operand))
```

Comment:
This is too loose for my taste. I'd like it to fail if it's given something that's already flattened.